File: 4sum.cpp
/* https://leetcode.com/problems/4sum-ii/ 
O(n) solution
a + b + c + d = 0 <=> a + b = -(c+d)
Store frequencies of sum of a+b and c+d in 2 dictionaries
*/

class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        int N = A.size();
        unordered_map<int, int> SumAB, SumCD;
        
        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < N; j++)
            {
                SumAB[A[i] + B[j]] ++;
                SumCD[C[i] + D[j]] ++;
            }
        }
        
        int count = 0;
        for (auto x : SumAB)
        {
            auto it = SumCD.find(x.first*-1);
            if (it != SumCD.end())
            {
                count += x.second * it->second;
            }
        }
        
        return count;
    }
};

File: 4sum.py
""" https://leetcode.com/problems/4sum-ii/ 
O(n^2) solution
"""

from collections import defaultdict

class Solution:
    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        n = len(A)
        count = 0
        SumAB= defaultdict(lambda: 0)
        SumCD = defaultdict(lambda: 0)
        
        for i in range(n):
            for j in range(n):
                SumAB[A[i] + B[j]] += 1
                SumCD[C[i] + D[j]] += 1
                
        for key1 in SumAB:
            if -key1 in SumCD:
                count += SumAB[key1] * SumCD[-key1]
                            
        return count
